rfcs/hse/0008-rework-cn-stats.md

```diff
@@ -0,0 +1,182 @@
+   headers, and others are stored in the kvset hblock.
+
+### Misleading statistics
+
+1. `key_stats.nvals` measures number of values associated with key.  It includes
+   tombstones but not prefix tombstones.
+1. `key_stats.tot_vlen`, `kblk_metrics.tot_val_bytes`,
+   `kvset_metrics.tot_val_bytes`: These measure the total on-media length of all
+   values associated with key/kblock/kvset.  It would be nice to have more
+   detail such as: length of vals in kblocks (VTYPE_IVAL) or maybe just lengths
+   of values by type (enum kmd_type).
+1. `kvset_metrics.num_keys` measures number of keys. It includes keys with
+   tombstones but not keys with prefix tombstones.
+1. `kvset_metrics.tot_key_bytes` measures sum of key lengths. In includes keys
+   with tombstones but not keys with prefix tombstones. It is also a wildly
+   inaccurate measure of on-media key length since it doesn't account for
+   longest common prefix elimination.  Nonetheless, it is used as if it were a
+   measure of on-media key length.
+
+## Solution
+
+### Overview
+
+1. Reduce the number of stat struct types.
+1. Simplify on-media storage of stats.
+1. Eliminate misleading stats.
+1. Eliminate the need to compute kvset level stats by iterating over all kblocks in kvset.
+
+### Details
+
+1. Eliminate `struct kblk_metrics` and `struct kvset_metrics` use `struct kvset_stats` more widely.
+   - Advantages:
+     - Eliminate the overhead of computing `struct kvset_metrics`.
+
+1. In `struct kvset_stats`, store `vlen` and `vcount` stats for each `enum
+   kmd_vtype` and eliminate special named fields such as `nptombs`.
+   - Advantages:
+     - Eliminates all problems identified in "Misleading Statistics".
+
+1. Store kvset-specific instances of the `one_stat_struct` in kvset hblocks
+   and remove related data them from CNDB.
+   - Advantages:
+     - Would not have to walk all kblocks to compute kvset level stats.
+     - Simplify CNDB.
+     - Simplify kvset open process.
+
+1. Store kblock-specific members of `struct kvset_stats` in kblock headers for
+   use during node split.  But in the provide access to it in memory as an
+   instance of `struct kvset_stats`.
+   - Advantages:
+       - Eliminate need for `struct kblk_metrics`
+     - Dead simple to compute kvset stats from kblock stats.
+
+1. Use `struct kvset_stats` with kvset builders instead of `struct key_stats`.
+   - Advantages:
+     - Eliminate need for `struct key_stats`.
+
+1. Replace `vused` with its inverse, `vgarbage`
+   - Advantages:
+     - A kv-compacted kvset will have `vgarbage == 0`, which is much easier to
+       work with than `vused` which has an unspecified non-zero value for nicely
+       compacted kvset (it equals the vblock size minus overhead such as headers
+       and padding).
+
+## Current stat struct definitions
+
+### struct key_stats
+Used by kvset builder to track statistics about a key's values during kvset
+construction.
+```
```

Comment:
add an empty newline prior to the triple backtick, and in your code blocks denote the language by adding `c` after the first triple backtick